src/language_features/texlab.rs

```diff
@@ -0,0 +1,95 @@
+use crate::context::Context;
+use crate::position::get_lsp_position;
+use crate::types::{EditorMeta, EditorParams};
+use crate::PositionParams;
+use lsp_types::request::Request;
+use lsp_types::TextDocumentIdentifier;
+use lsp_types::TextDocumentPositionParams;
+use lsp_types::Url;
+use serde::{Deserialize, Serialize};
+
+pub enum ForwardSearch {}
+
+impl Request for ForwardSearch {
+    type Params = TextDocumentPositionParams;
+    type Result = ForwardSearchResult;
+    const METHOD: &'static str = "textDocument/forwardSearch";
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ForwardSearchResult {
+    status: ForwardSearchStatus,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub enum ForwardSearchStatus {
+    Success = 0,
+    Error = 1,
+    Failure = 2,
+    Unconfigured = 3,
+}
+
+pub fn forward_search(meta: EditorMeta, params: EditorParams, ctx: &mut Context) {
+    let params = PositionParams::deserialize(params).unwrap();
+    let req_params = TextDocumentPositionParams {
+        text_document: TextDocumentIdentifier {
+            uri: Url::from_file_path(&meta.buffile).unwrap(),
+        },
+        position: get_lsp_position(&meta.buffile, &params.position, ctx).unwrap(),
+    };
+    ctx.call::<ForwardSearch, _>(meta, req_params, move |ctx, meta, response| {
+        forward_search_response(meta, response, ctx)
+    });
+}
+
+pub fn forward_search_response(meta: EditorMeta, result: ForwardSearchResult, ctx: &mut Context) {
+    let command = format!("echo {:?}", result);
+    ctx.exec(meta, command);
+}
+
+pub enum Build {}
+
+impl Request for Build {
+    type Params = BuildTextDocumentParams;
+    type Result = BuildResult;
+    const METHOD: &'static str = "textDocument/build";
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct BuildTextDocumentParams {
+    text_document: TextDocumentIdentifier,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct BuildResult {
+    status: BuildStatus,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+pub enum BuildStatus {
+    Success = 0,
+    Error = 1,
+    Failure = 2,
+    Cancelled = 3,
+}
+
+pub fn build(meta: EditorMeta, _params: EditorParams, ctx: &mut Context) {
+    let req_params = BuildTextDocumentParams {
+        text_document: TextDocumentIdentifier {
+            uri: Url::from_file_path(&meta.buffile).unwrap(),
+        },
+    };
+    ctx.call::<Build, _>(meta, req_params, move |ctx, meta, response| {
+        build_response(meta, response, ctx)
+    });
+}
+
+pub fn build_response(meta: EditorMeta, result: BuildResult, ctx: &mut Context) {
+    let command = format!("echo {:?}", result);
```

Comment:
This command is executed when texlab indicates a build result.

Is the echo of the Build Statue here correct?

Reply:
The problem with `ctx.exec(meta, "echo something")` is that it runs without a (Kakoune) client in the context.
It's equivalent to running `echo echo something | kak -p my-session`.

Luckily, since we know the Kakoune client who made the request. So something like (untested)

```rust
let command = format!("evaluate-commands -client {} %{ echo the-status }", meta.client.unwrap());
ctx.exec(meta, command);
```

should work. I also like to use `echo -debug` for debugging (that one does not need a client context).

Reply:
The Problem was that I did not pass the client to the meta when calling build in `lsp.kak`. `kak-lsp` uses `eval -client <client> <cmd>` automatically when invoking `ctx.exec(meta,command)` when the meta specifies a client, so a change in lsp.kak was sufficient.

Reply:
Right, it's always a good idea to check the `*debug*` buffer for error messages